src/base/QXmppElement.cpp

```diff
@@ -123,10 +123,13 @@ QXmppElement::~QXmppElement()
 
 QXmppElement &QXmppElement::operator=(const QXmppElement &other)
 {
-    other.d->counter.ref();
-    if (!d->counter.deref())
-        delete d;
-    d = other.d;
+    if (this != &other) // self-assignment check
```

Comment:
> This check doesn't actually affect correctness, right?

Yep.

> The performance improvements might be marginal

Are you sure? This looks quite long:
https://github.com/qxmpp-project/qxmpp/blob/034c318/src/base/QXmppElement.cpp#L53